bokehjs/src/lib/models/tools/gestures/range_tool.ts

```diff
@@ -46,6 +54,161 @@ export class RangeToolView extends ToolView {
+      top: has_y ? lrtb.top.invert(top) : this.model.nodes.top,
+      bottom: has_y ? lrtb.bottom.invert(bottom) : this.model.nodes.bottom,
+    }
+  }
+
+  protected _compute_limits(curr_point: [number, number]): [[number, number], [number, number]] {
+    const dims = (() => {
+      const {x_range, y_range} = this.model
+      const has_x = x_range != null
+      const has_y = y_range != null
+
+      if (has_x && has_y) {
+        return "both"
+      } else if (has_x) {
+        return "width"
+      } else if (has_y) {
+        return "height"
+      } else {
+        unreachable()
+      }
+    })()
+
+    assert(this._base_point != null)
+    let base_point = this._base_point
+    if (this.model.overlay.symmetric) {
+      const [cx, cy] = base_point
+      const [dx, dy] = curr_point
+      base_point = [cx - (dx - cx), cy - (dy - cy)]
+    }
+
+    const {frame} = this.plot_view
+    return this.model._get_dim_limits(base_point, curr_point, frame, dims)
+  }
+
+  protected _base_point: [number, number] | null
+
+  override _tap(ev: TapEvent): void {
+    assert(this.model.select_gesture == "tap")
+
+    const {sx, sy} = ev
+    const {frame} = this.plot_view
+    if (!frame.bbox.contains(sx, sy)) {
+      return
+    }
+
+    if (this._base_point == null) {
+      this._base_point = [sx, sy]
+    } else {
+      this._update_overlay(sx, sy)
+      this._base_point = null
+    }
+  }
+
+  override _move(ev: MoveEvent): void {
+    if (this._base_point != null && this.model.select_gesture == "tap") {
+      const {sx, sy} = ev
+      this._update_overlay(sx, sy)
+    }
+  }
+
+  override _pan_start(ev: PanEvent): void {
+    assert(this.model.select_gesture == "pan")
+    assert(this._base_point == null)
+
+    const {sx, sy} = ev
+    const {frame} = this.plot_view
+    if (!frame.bbox.contains(sx, sy)) {
+      return
+    }
+
+    this._base_point = [sx, sy]
+  }
+
+  protected _update_overlay(sx: number, sy: number): void {
+    const [sxlim, sylim] = this._compute_limits([sx, sy])
+    const [[left, right], [top, bottom]] = [sxlim, sylim]
+    this.model.overlay.update(this._invert_lrtb({left, right, top, bottom}))
+    this.model.update_ranges_from_overlay()
+  }
+
+  override _pan(ev: PanEvent): void {
+    if (this._base_point == null) {
+      return
+    }
+
+    const {sx, sy} = ev
+    this._update_overlay(sx, sy)
+  }
+
+  override _pan_end(ev: PanEvent): void {
+    if (this._base_point == null) {
+      return
+    }
+
+    const {sx, sy} = ev
+    this._update_overlay(sx, sy)
+
+    this._base_point = null
+  }
+
+  protected get _is_selecting(): boolean {
+    return this._base_point != null
+  }
+
+  protected _stop(): void {
+    this._base_point = null
+  }
+
+  override _keyup(ev: KeyEvent): void {
+    if (!this.model.active) {
+      return
+    }
+
+    if (ev.key == "Escape") {
+      if (this._is_selecting) {
+        this._stop()
+        return
+      }
+    }
```

Comment:
This seems like it could be a bit more compat
```suggestion
    if (this.model.active) {
      if (ev.key == "Escape" && this._is_selecting) {
        this._stop()
      }
    }
```